lib/eqaf.ml

```diff
@@ -1,11 +1,21 @@
-external unsafe_string_get : string -> int -> int = "%string_unsafe_get"
+external unsafe_string_get : string -> int -> int = "%string_unsafe_get" [@@noalloc]
+external unsafe_string_get_64 : string -> int -> int64 = "%caml_string_get64u"
 
 let equal a b =
-  let ln = String.length a in
-  if ln = String.length b then (
-    let rt = ref 0 in
-    for i = 0 to ln - 1 do
-      rt := !rt lor (unsafe_string_get a i lxor unsafe_string_get b i)
-    done ;
-    !rt = 0 )
-  else false
+  let len = min (String.length a) (String.length b) in
+  let res = ref 0L in
+  for i = 0 to pred (len / 8) do
+    res :=
+      Int64.logor !res
+        (Int64.logxor
+           (unsafe_string_get_64 a (i * 8))
+           (unsafe_string_get_64 a (i * 8)))
+  done ;
+  for i = len - (len mod 8) to pred len do
+    res :=
+      Int64.logor !res
```

Comment:
Why do you use Int64 instead eg Nativeint here? Won't the 64bit expansion be slower on 32bit?

Reply:
I did a benchmark for you to compare `eqnt` which use the `Native` module and `eqbg`/`eqaf` which uses `Int64` and I got this result on my server (no virtualization):

```sh
┌─────────────────┬─────────────┬────────────┐
│ Name            │    Time/Run │ Percentage │
├─────────────────┼─────────────┼────────────┤
│ diff eqaf:0     │     21.48ns │      0.16% │
│ diff eqaf:10    │     25.27ns │      0.18% │
│ diff eqaf:100   │     42.39ns │      0.31% │
│ diff eqaf:1000  │    186.02ns │      1.36% │
│ diff eqaf:2000  │    338.12ns │      2.47% │
│ diff eqaf:4000  │    642.03ns │      4.68% │
│ diff eqaf:8000  │  1_249.76ns │      9.12% │
│ diff eqaf:10000 │  1_553.53ns │     11.34% │
│ same eqaf:0     │     21.49ns │      0.16% │
│ same eqaf:10    │     25.25ns │      0.18% │
│ same eqaf:100   │     42.38ns │      0.31% │
│ same eqaf:1000  │    186.04ns │      1.36% │
│ same eqaf:2000  │    338.02ns │      2.47% │
│ same eqaf:4000  │    642.03ns │      4.68% │
│ same eqaf:8000  │  1_249.83ns │      9.12% │
│ same eqaf:10000 │  1_553.38ns │     11.33% │
│ diff eqst:0     │      7.75ns │      0.06% │
│ diff eqst:10    │      7.75ns │      0.06% │
│ diff eqst:100   │      7.75ns │      0.06% │
│ diff eqst:1000  │      7.75ns │      0.06% │
│ diff eqst:2000  │      7.76ns │      0.06% │
│ diff eqst:4000  │      7.76ns │      0.06% │
│ diff eqst:8000  │      7.75ns │      0.06% │
│ diff eqst:10000 │      7.75ns │      0.06% │
│ same eqst:0     │      7.75ns │      0.06% │
│ same eqst:10    │      8.66ns │      0.06% │
│ same eqst:100   │     25.84ns │      0.19% │
│ same eqst:1000  │     95.72ns │      0.70% │
│ same eqst:2000  │    156.74ns │      1.14% │
│ same eqst:4000  │    277.51ns │      2.02% │
│ same eqst:8000  │    565.09ns │      4.12% │
│ same eqst:10000 │    643.67ns │      4.70% │
│ diff eqnt:0     │     18.33ns │      0.13% │
│ diff eqnt:10    │     31.00ns │      0.23% │
│ diff eqnt:100   │    161.82ns │      1.18% │
│ diff eqnt:1000  │  1_392.18ns │     10.16% │
│ diff eqnt:2000  │  2_759.05ns │     20.13% │
│ diff eqnt:4000  │  5_496.20ns │     40.10% │
│ diff eqnt:8000  │ 10_963.76ns │     80.00% │
│ diff eqnt:10000 │ 13_693.26ns │     99.91% │
│ same eqnt:0     │     18.69ns │      0.14% │
│ same eqnt:10    │     30.99ns │      0.23% │
│ same eqnt:100   │    162.40ns │      1.18% │
│ same eqnt:1000  │  1_392.54ns │     10.16% │
│ same eqnt:2000  │  2_759.47ns │     20.13% │
│ same eqnt:4000  │  5_494.26ns │     40.09% │
│ same eqnt:8000  │ 10_962.34ns │     79.99% │
│ same eqnt:10000 │ 13_702.32ns │     99.98% │
│ diff eqbg:0     │     20.37ns │      0.15% │
│ diff eqbg:10    │     24.71ns │      0.18% │
│ diff eqbg:100   │     41.03ns │      0.30% │
│ diff eqbg:1000  │    183.84ns │      1.34% │
│ diff eqbg:2000  │    335.77ns │      2.45% │
│ diff eqbg:4000  │    639.62ns │      4.67% │
│ diff eqbg:8000  │  1_246.88ns │      9.10% │
│ diff eqbg:10000 │  1_551.20ns │     11.32% │
│ same eqbg:0     │     20.36ns │      0.15% │
│ same eqbg:10    │     24.71ns │      0.18% │
│ same eqbg:100   │     41.03ns │      0.30% │
│ same eqbg:1000  │    183.89ns │      1.34% │
│ same eqbg:2000  │    335.36ns │      2.45% │
│ same eqbg:4000  │    639.38ns │      4.67% │
│ same eqbg:8000  │  1_247.10ns │      9.10% │
│ same eqbg:10000 │  1_551.12ns │     11.32% │
│ diff eqml:0     │      5.47ns │      0.04% │
│ diff eqml:10    │     19.45ns │      0.14% │
│ diff eqml:100   │    151.40ns │      1.10% │
│ diff eqml:1000  │  1_382.76ns │     10.09% │
│ diff eqml:2000  │  2_752.33ns │     20.08% │
│ diff eqml:4000  │  5_490.57ns │     40.06% │
│ diff eqml:8000  │ 10_969.67ns │     80.04% │
│ diff eqml:10000 │ 13_701.77ns │     99.98% │
│ same eqml:0     │      5.93ns │      0.04% │
│ same eqml:10    │     19.45ns │      0.14% │
│ same eqml:100   │    152.04ns │      1.11% │
│ same eqml:1000  │  1_382.86ns │     10.09% │
│ same eqml:2000  │  2_751.57ns │     20.08% │
│ same eqml:4000  │  5_487.31ns │     40.04% │
│ same eqml:8000  │ 10_969.40ns │     80.04% │
│ same eqml:10000 │ 13_705.11ns │    100.00% │
└─────────────────┴─────────────┴────────────┘
```

As you can see `eqbg` is more fast than `eqnt`: `1551ns < 13702ns` for `string` with a length `= 10000`, so I think, we will keep `eqbg`/`eqaf` of this PR

Reply:
Does your server run on a 32bit processor?

Reply:
`amd64`, I did not have an access to a 32-bits machine :(